hs-bindgen-libclang/src/HsBindgen/Clang/Internal/Results.hs

```diff
@@ -104,28 +106,35 @@ ensureNotInRange = ensureOn conv (not . simpleEnumInRange)
     conv = coerceSimpleEnum . fromIntegral
 
 {-------------------------------------------------------------------------------
-  Check wrapper result
+  Clang version error
 -------------------------------------------------------------------------------}
 
-data Unsupported = Unsupported (SimpleEnum WrapperResult) Backtrace
+-- | Clang version requirement
+data ClangVersionRequirement =
+    ClangTooOld     -- ^ Clang is too old (older than version 11)
+  | Clang11Required -- ^ We need at least Clang version 11
+  | Clang12Required -- ^ We need at least Clang version 12
+  | Clang13Required -- ^ We need at least Clang version 13
+  | Clang14Required -- ^ We need at least Clang version 14
+  | Clang15Required -- ^ We need at least Clang version 15
+  | Clang16Required -- ^ We need at least Clang version 16
+  | Clang17Required -- ^ We need at least Clang version 17
+  | Clang18Required -- ^ We need at least Clang version 18
+  deriving stock (Show, Eq, Enum, Bounded, Generic)
+
+-- | The version of Clang being used is not supported
+data ClangVersionError = ClangVersionError ClangVersionRequirement Backtrace
```

Comment:
Can we re-use `ClangVersion` instead of re-enumerating the types in `ClangVersionRequirement`? We cannot make more fine-grained distinctions than `ClangVersion`.

Reply:
Yes, that should not be a problem.  I rebased to include merges to `main` and made the change.

Details:

(Some) `ClangVersion` constructors map to more than one Clang version, but not being able to specify requirements for intermediate versions should never be a problem because `CINDEX_VERSION_MINOR` is bumped on API additions.

We use `Show` instances when displaying exceptions, so we can simply wrap a `ClangVersion` value with a type that indicates the requirement, being sure to derive with the `stock` strategy.  Example: `Requires Clang17_or_18_or_19` would indicate that Clang 17 or newer is required.

I went ahead and used `CallFailed` for this as well.  The implementation is in `HsBindgen.Clang.Version` to avoid an import cycle.  I think that this (exposed module) is a good place for it, as users may want to use `requireClangVersion` themselves (when using a higher-level API).

Reply:
Ok, this looks great now, thank you :) And indeed, `ClangVersion` isn't as granular as one would like, but it is as granular as we can test for, so I think it makes sense to reuse it here.

By the way, thank you for sticking to local coding style, that is much appreciated!